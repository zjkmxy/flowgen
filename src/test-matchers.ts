import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
