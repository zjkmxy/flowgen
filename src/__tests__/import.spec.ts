import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle dynamic imports", async () => {
  const ts = `
// whole module
type A = import('react');

// type alias inside module
type B = import('react').ReactNode;

// generic type alias inside module, with type arguments
type C = import('react').ComponentType<{}>;

// class inside module
type D = import('zlib').Zlib;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle imports from odd names", async () => {
  const ts = `
type A = import('..');
type B = import('@!-/./');
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  // expect(result).toBeValidFlowTypeDeclarations(); // would need actual modules at those names
});

it("should handle import nested in type arguments of import", async () => {
  // In other words, test that our visitor for this feature didn't forget to
  // visit the type arguments in the case where it's rewriting something.
  const ts = `
type A = import("react").ComponentType<import("react").ComponentType<any>>;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle import in an imported file", async () => {
  // There once was a bug where transformers wouldn't get run on the second
  // (or later) file in a list like this.  Test that the transformer
  // implementing this feature does indeed run there.
  const results = compiler.compileDefinitionFiles(
    [
      "src/__tests__/snippet/import-import-type.ts",
      "src/__tests__/snippet/import-type.ts",
    ],
    { quiet: false },
  );
  for (const result of results) {
    expect(await beautify(result[1])).toMatchSnapshot(result[0]);
  }
});
