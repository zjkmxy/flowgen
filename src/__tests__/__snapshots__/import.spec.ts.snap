// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle dynamic imports 1`] = `
"import * as $Flowgen$Import$react from "react";
import * as $Flowgen$Import$zlib from "zlib";
declare type A = typeof $Flowgen$Import$react;
declare type B = $Flowgen$Import$react.Node;
declare type C = $Flowgen$Import$react.ComponentType<{}>;
declare type D = $Flowgen$Import$zlib.Zlib;
"
`;

exports[`should handle import in an imported file: src/__tests__/snippet/import-import-type.ts 1`] = `
"declare export { R } from "./import-type";
"
`;

exports[`should handle import in an imported file: src/__tests__/snippet/import-type.ts 1`] = `
"import * as $Flowgen$Import$react from "react";
declare export var R: $Flowgen$Import$react.RefAttributes<any>;
"
`;

exports[`should handle import nested in type arguments of import 1`] = `
"import * as $Flowgen$Import$react from "react";
declare type A = $Flowgen$Import$react.ComponentType<
  $Flowgen$Import$react.ComponentType<any>,
>;
"
`;

exports[`should handle imports from odd names 1`] = `
"import * as $Flowgen$Import$_2e__2e_ from "..";
import * as $Flowgen$Import$_40__21__2d__2f__2e__2f_ from "@!-/./";
declare type A = typeof $Flowgen$Import$_2e__2e_;
declare type B = typeof $Flowgen$Import$_40__21__2d__2f__2e__2f_;
"
`;
